src/auth/authenticator/tls/authenticator.go

```diff
@@ -0,0 +1,56 @@
+package tls
+
+import (
+	"bytes"
+	"crypto/x509"
+	"net/http"
+
+	"github.com/consensys/quorum-key-manager/pkg/errors"
+	"github.com/consensys/quorum-key-manager/src/auth/types"
+)
+
+const AuthMode = "Tls"
+
+type Authenticator struct {
+	Certificates []*x509.Certificate
+}
+
+func NewAuthenticator(cfg *Config) (*Authenticator, error) {
+	if len(cfg.Certificates) == 0 {
+		return nil, nil
+	}
+
+	auth := &Authenticator{Certificates: cfg.Certificates}
+
+	return auth, nil
+}
+
+// Authenticate checks certs and retrieve user Info
+// CN field -> Username
+// Organization -> Groups
+func (authenticator Authenticator) Authenticate(req *http.Request) (*types.UserInfo, error) {
+	// extract Certificate info from request if any
+	if len(req.TLS.PeerCertificates) == 0 {
+		return types.AnonymousUser, nil
```

Comment:
you should return nil, nil and then AuthenticathorMid will take care of appending `AnnonymousUser` groups to the UserInfo attached in the context

Reply:
Ok, I have changed this. But in that case I am not sure that the duty of the upper layer is to guess what user it is dealing with. Responsability of Authenticate is to return useInfo. Here we should probably consider always returning a proper user instance and let the upper layer deal with it.

Reply:
that would work in case of having a single auth alternatives. In meanwhile we are supporting several ones it wont work since we need a parent auth components to decide the final outcome. What else would you propose here?